.github/workflows/ci-test.yml

```diff
@@ -0,0 +1,143 @@
+name: ci-test
+
+on:
+  push:
+    branches: [auto, gha-test-*]
+  pull_request:
+    branches: [main, master, release-*, gha-test-*]
+
+jobs:
+  prepare:
+    runs-on: ubuntu-latest
+    continue-on-error: false
+    outputs:
+      changes-target-branch: ${{ steps.changes.outputs.changes-target-branch }}
+      changes-base-githash: ${{ steps.changes.outputs.changes-base-githash }}
+      changes-pull-request-number: ${{ steps.changes.outputs.changes-pull-request-number }}
+      test-rust: ${{ steps.rust-changes.outputs.changes-found }}
+      test-dev-setup: ${{ steps.dev-setup-sh-changes.outputs.changes-found }}
+      test-website-build: ${{ steps.website-changes.outputs.changes-found }}
+      test-non-rust-lint: ${{ steps.non-rust-lint-changes.outputs.changes-found }}
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 #get all the history!!!
+      - id: changes
+        name: determine changes
+        uses: ./.github/actions/changes
+        with:
+          workflow-file: ci-test.yml
+      - id: rust-changes
+        name: find rust/cargo changes.
+        uses: ./.github/actions/matches
+        with:
+          pattern: '**.rs\|Cargo.toml\|Cargo.lock'
+      - id: non-rust-lint-changes
+        name: find shell/dockerfile changes
+        uses: ./.github/actions/matches
+        with:
+          pattern: '**.Dockerfile\|**.sh'
+      - id: dev-setup-sh-changes
+        name: find dev-setup.sh/base docker image changes
+        uses: ./.github/actions/matches
+        with:
+          pattern: '.github/workflows/ci-publish-base-image.yml\|docker/ci/**/Dockerfile\|scripts/dev_setup.sh'
+      - id: website-changes
+        name: find website changes.
+        uses: ./.github/actions/matches
+        with:
+          pattern: "documentation/**"
+
+  dev-setup-sh-test:
+    runs-on: ubuntu-latest-xl
+    continue-on-error: false
+    needs: prepare
+    if: ${{ needs.prepare.outputs.test-dev-setup == 'true' }}
```

Comment:
and here's why we have the prepare section.   Only test the dev-setup.sh script in various docker environments if relevant files have changed, otherwise dont even start up the machines.